src/manager/evm.rs

```diff
@@ -0,0 +1,379 @@
+        _from: Address,
+        _to_add: Vec<Address>,
+    ) -> Result<()> {
+        todo!()
+    }
+
+    /// Send value between two addresses in a subnet
+    async fn send_value(&self, _from: Address, _to: Address, _amount: TokenAmount) -> Result<()> {
+        todo!()
+    }
+
+    async fn wallet_new(&self, _key_type: WalletKeyType) -> Result<Address> {
+        todo!()
+    }
+
+    async fn wallet_list(&self) -> Result<Vec<Address>> {
+        todo!()
+    }
+
+    async fn wallet_balance(&self, _address: &Address) -> Result<TokenAmount> {
+        todo!()
+    }
+
+    async fn last_topdown_executed(&self, _gateway_addr: &Address) -> Result<ChainEpoch> {
+        todo!()
+    }
+
+    async fn list_checkpoints(
+        &self,
+        _subnet_id: SubnetID,
+        _from_epoch: ChainEpoch,
+        _to_epoch: ChainEpoch,
+    ) -> Result<Vec<BottomUpCheckpoint>> {
+        todo!()
+    }
+}
+
+impl<M: Middleware + Send + Sync> EthSubnetManager<M> {
+    pub fn new(
+        eth_client: Arc<M>,
+        gateway_contract: Gateway<Arc<M>>,
+        registry_contract: SubnetRegistry<Arc<M>>,
+    ) -> Self {
+        Self {
+            eth_client,
+            gateway_contract,
+            registry_contract,
+        }
+    }
+
+    pub fn ensure_same_gateway(&self, gateway: &Address) -> anyhow::Result<()> {
+        let evm_gateway_addr = payload_to_evm_address(gateway.payload())?;
+        if evm_gateway_addr != (*self.gateway_contract).address() {
+            Err(anyhow!("Gateway address not matching with config"))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl EthSubnetManager<MiddlewareImpl> {
+    pub fn from_subnet(subnet: &Subnet) -> Result<Self> {
+        let url = subnet.rpc_http().clone();
+        let auth_token = subnet.auth_token();
+
+        let config = if let SubnetConfig::Evm(config) = &subnet.config {
+            config
+        } else {
+            return Err(anyhow!("not evm config"));
+        };
+
+        let provider = if auth_token.is_some() {
+            Http::new_with_auth(url, Authorization::Bearer(auth_token.unwrap()))?
+        } else {
+            Http::new(url)
+        };
+
+        let provider = Provider::new(provider);
+        let wallet = config.private_key.parse::<LocalWallet>()?;
+        let wallet = wallet.with_chain_id(subnet.id.chain_id());
+
+        let gateway_address = payload_to_evm_address(config.gateway_addr.payload())?;
+        let registry_address = payload_to_evm_address(config.registry_addr.payload())?;
+
+        let signer = Arc::new(SignerMiddleware::new(provider, wallet));
+        let gateway_contract = Gateway::new(gateway_address, Arc::new(signer.clone()));
+        let evm_registry_contract = SubnetRegistry::new(registry_address, Arc::new(signer.clone()));
+
+        Ok(Self::new(signer, gateway_contract, evm_registry_contract))
+    }
+}
+
+/// Convert the ipc SubnetID type to an evm address. It extracts the last address from the Subnet id
```

Comment:
Why can't we make these two functions just one? You remove all of the childrens and then take the ones you are interested in in the relevant function? I am not a fan of functions that have a single character as the only difference in their name :)